fake_spectra/parallel.py

```diff
@@ -0,0 +1,85 @@
+import fake_spectra
+#print("fake_spectra imported")
+from fake_spectra.randspectra import RandSpectra
+#print("haloassigned_spectra imported")
+from fake_spectra.plot_spectra import PlottingSpectra
+#print("PlottingSpectra imported")
+from mpi4py import MPI
+import numpy as np
+
+comm = MPI.COMM_WORLD
+rank = comm.Get_rank()
+size = comm.Get_size()
+
+numlos = 1000
+ndla = 1000
+thresh_t = 10**20.3
+### the array to store added column density in
+
+
+
+if rank ==0 :
+    
+    found = 0
+
+    while (found < ndla) :
+
+        col_den_added = np.zeros(shape=(numlos,))
+
+        ### Recieve col_den from all other ranks and add them together
+        comm.Reduce(np.zeros(shape=(numlos,), dtype='d'), col_den_added, op = MPI.SUM, root = 0)
```

Comment:
This function needs to be called on all ranks to be effective. When the function is called, what happens is something like this:
if rank == root:
MPI_Receive(data, all_ranks)
sum(data)
else:
MPI_Send(data, root)

Because this is not a shared memory system the data cannot be simply read by the master rank: it has to be sent from the other ranks as well. A good rule for MPI is that all collective operations have to be called by every rank in the communicator.

This is because otherwise the code doesn't know at what point in execution to read data: these processes are not guaranteed to run in any particular order. So the Send call means "get here and then send the message".

Reply:
> This function needs to be called on all ranks to be effective. When the function is called, what happens is something like this:
> if rank == root:
> MPI_Receive(data, all_ranks)
> sum(data)
> else:
> MPI_Send(data, root)
> 
> Because this is not a shared memory system the data cannot be simply read by the master rank: it has to be sent from the other ranks as well. A good rule for MPI is that all collective operations have to be called by every rank in the communicator.
> 
> This is because otherwise the code doesn't know at what point in execution to read data: these processes are not guaranteed to run in any particular order. So the Send call means "get here and then send the message".

In replace_not_DLA  in spectra.py , I have called comm.Reduce(colden, colden_added, ...). So, after calculating the column densities in each rank they sent back the column densities to rank = 0. Then rank =0 returns all indices which are above threshold to all labor ranks, then they are able to replace those ranks.
So, do you mean the comm.Reduce() should be called in other places than replace_not_DLA ?